### ISSUE TITLE
Firehose delivery missing?

### ISSUE THREAD

**Issue by MacksMind** [awslabs/lambda-streams-to-firehose#47]
I'm using dynamodb-continuous-backup for the first time, and have an issue I don't understand. I was monitoring the S3 bucket to test something else when I realized one of the inserts never arrived. DynamoDB shows 6 entries for a given partition key, but only 5 made it into S3. CloudWatch show an appropriate entry "2017-05-10T19:19:28.861Z	14502b25-707b-4ecc-9697-4321b747b9c1	Event forwarding complete. Forwarded 1 batches comprising 1 records to Firehose production_tracking", but the data never arrived.

So for 6 inserts in a 20 hours period, the 3rd one is missing, and I didn't change anything except the data during this time. The CloudWatch log is the only evidence I can find.

Any ideas?

**Comment by IanMeyers** [comment]
Gosh - sorry to hear that this is the case. This isn't something that I've seen happen before, so we'll have to do some investigation. Would be interesting to know a bit about the timing of the missing record from s3. Was it much later than the other 5, or anything else different about it? What are the buffer settings for the firehose delivery stream? Could it have been delivered to a different s3 prefix, for example?

**Comment by MacksMind** [comment]
The records in this table have timestamps, so I have:

* 2017-05-10T19:09:39.486Z - ok
* 2017-05-10T19:19:25.870Z - missing
* 2017-05-10T19:24:44.400Z - ok

Nothing else was special - in fact the 3 records are almost identical. Here's an example record:
```
{
  "latitude": 39.9862,
  "longitude": -86.0988,
  "professional_id": "82b206c3-c3bf-4afb-96a9-047d8c36a7db",
  "timestamp": "2017-05-10T19:09:39.486Z"
}
```

And the corresponding S3 entry (formatted):
```
{
  "Keys": {
    "professional_id":{
      "S":"82b206c3-c3bf-4afb-96a9-047d8c36a7db"
    },
    "timestamp":{
      "S":"2017-05-10T19:09:39.486Z"
    }
  },
  "NewImage":{
    "latitude":{
      "N":"39.9862"
    },
    "professional_id":{
      "S":"82b206c3-c3bf-4afb-96a9-047d8c36a7db"
    },
    "longitude":{
      "N":"-86.0988"
    },
    "timestamp":{
      "S":"2017-05-10T19:09:39.486Z"
    }
  },
  "SequenceNumber":"22876800000000007920067356",
  "SizeBytes":194,
  "ApproximateCreationDateTime":1494443340,
  "eventName":"INSERT"
}
```

The buffer settings are 128MB and 60 seconds. The S3 bucket and prefix haven't changed since provisioning by dynamodb-continuous-backup. Also I notice an error logging setting, but it's set to Disabled.

**Comment by kyleaa** [comment]
I've had a table set up using this lambda for the past 9 months, and just noticed that out of 43 million records we are missing about 1k - they appear to have just been dropped. There is no pattern or time correlation to them. They're not in S3 so it's theoretically possible for Kinesis or Dynamo Streams to be the culprit. I'm going to try to enable debugging in the lambda and see if I can get another occurrence to narrow it down more. I had been running on 1.3.5 and just upgraded to 1.4.5.

**Comment by DaichiUeura** [comment]
If v1.4.5 resolves the issue, https://github.com/awslabs/lambda-streams-to-firehose/issues/37 could be the cause.

**Comment by benoittgt** [comment]
@kyleaa I confirm @DaichiUeura  no more missing events since the update.

**Comment by hernangarcia** [comment]
Hello all!

I am wondering if anyone is still able to reproduce this behaviour. It is happening to me very often with no heavy loads. 

I have a DynamoDB table stream enabled mapped to a Firehose with the same table name that should write to a RedShift table.

I can see the lambda function is retrieving the new entry but that's it. There's not an S3 file with it much less any log in Kinesis or entry in RedShift.

**My config**
S3 buffer size (MB) = 5
S3 buffer interval (sec) = 60

I have multiple dynamoDB streams triggering the same  LambdaStreamToFirehose v1.5 function.

Thanks in advance.

**Comment by benoittgt** [comment]
@hernangarcia I don't have anymore issues on my side. The patch from @DaichiUeura fixed the issue.

**Comment by DaichiUeura** [comment]
What happen if you use v1.4.5? (https://github.com/awslabs/lambda-streams-to-firehose/issues/53 might be related)

**Comment by hernangarcia** [comment]
Thanks a lot @benoittgt. I added the patch from @DaichiUeura . I am monitoring again, let you guys know. If it does not work, I'll downgrade to 1.4.5.

Thanks!!!

**Comment by benoittgt** [comment]
Great news. Keep us updated!

**Comment by rhamnett** [comment]
Fix is available at issue #53
